File: website/functions/src/models/round.ts
import {Card} from './card';
import {Hand, Hands} from './hand';
import {Players} from './player';

export class Round {
  constructor(
    public discardPile: Card[] = [],
    public playersHand: Hands = new Hands(),
  ) {}

  static new(players: Players): Round {
    const cards = Card.all().sort(() => Math.random() - 0.5);
    const firstCard = cards.pop();

    const hands: Record<string, Hand> = {};

    for (const player of players.list) {
      hands[player.id] = new Hand(
          [],
          [],
          0,
      );
    }

    while (cards.length >= players.length) {
      for (const player of players.list) {
        const hand = hands[player.id];

        if (hand.revealedPile.length < 3) {
          hand.revealedPile.push(cards.pop()!);
        } else {
          hand.hiddenPile.push(cards.pop()!);
        }
      }
    }

    return new Round(
        [firstCard!],
        new Hands(hands),
    );
  }

  static parse(data: any): Round {
    if (data) {
      return new Round(
          Card.parseList(data['discardPile']),
          Hands.parse(data['playersHand']),
      );
    } else {
      return new Round();
    }
  }

  public get lastCard() {
    return this.discardPile[this.discardPile.length - 1];
  }

  public json() {
    return {
      discardPile: this.discardPile.map((c) => c.json()),
      playersHand: this.playersHand.json(),
    };
  }
}

File: server/lib/domain/models/round.dart
import 'dart:math';
import 'package:undervoltage/domain/json/game/json_round.dart';
import 'package:undervoltage/domain/models/card.dart';
import 'package:undervoltage/domain/models/hand.dart';
import 'package:undervoltage/domain/models/player.dart';
import 'package:undervoltage/utils/logger.dart';

class Round {
  final List<Card> discardPile;
  final Map<String, Hand> playersHand;

  const Round({
    required this.discardPile,
    required this.playersHand,
  });

  factory Round.create(List<Player> players) {
    final List<Card> cards = Card.all();
    cards.shuffle(Random());

    final Card firstCard = cards.removeAt(0);

    final Map<String, Hand> hands = {};

    for (final Player player in players) {
      final List<Card> hiddenPile = [];
      final List<Card> revealedPile = [];

      hands[player.id] = Hand(
        hiddenPile: hiddenPile,
        revealedPile: revealedPile,
        faults: 0,
      );
    }

    while (cards.length >= players.length) {
      for (final Player player in players) {
        final Hand hand = hands[player.id]!;
        final Card card = cards.removeAt(0);

        if (hand.revealedSize < 3) {
          hand.addRevealed(card);
        } else {
          hand.addHidden(card);
        }
      }
    }

    return Round(
      discardPile: [firstCard],
      playersHand: hands,
    );
  }

  JsonRound get json => JsonRound(
        discardPile: discardPile.map((e) => e.json).toList(),
        playersHand: playersHand.map((key, value) => MapEntry(key, value.json)),
      );

  Card get topCard => discardPile.last;

  void addCard(Card card) => discardPile.add(card);

  Hand playerHand(String playerId) => playersHand[playerId]!;

  bool hasHandFor(String playerId) => playersHand.containsKey(playerId);

  bool get isBlocked => discardPile.isNotEmpty && _allPlayersBlocked(topCard);

  bool _allPlayersBlocked(Card topCard) {
    for (final Hand hand in playersHand.values) {
      if (!hand.isBlocked(topCard)) {
        return false;
      }
    }

    return playersHand.isNotEmpty;
  }

  void unblock() {
    int limit = discardPile.length;

    while (isBlocked && (limit > 0)) {
      final Card bottom = discardPile.removeAt(0);
      discardPile.add(bottom);
      Logger.info('Moved card: $bottom to top of discard pile');
      limit--;
    }

    Logger.info('Round no longer blocked');
  }
}
